spec/concepts/jwt_satisfaction_survey/operation/create_spec.rb

```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe JwtSatisfactionSurvey::Operation::Create do
+  subject(:operation) { described_class }
+
+  before do
+    expect(JwtApiEntreprise).to receive(:seven_days_ago_created_tokens).and_call_original
+  end
```

Comment:
On a pas besoin de connaitre les rouages internes dans le cadre de ces specs. Si on devait mock une dépendance externe oui mais ce n'est pas le cas en l'occurence. 

On pourrait si on voulait résoudre des problèmes de perfs d'exécution des tests : stub le modèle `JwtApiEntreprise` et tester le comportement de l'opération en fonction de son retour. Et tester aussi `JwtApiEntreprise` en isolation pour s'assurer que la méthode `seven_days_ago_created_tokens` fait bien ce qu'on veut. 

Là en l'occurence c'est sur le comportement de l'opération que l'on doit porter attention donc 👍🏼 * créer plusieurs jetons en base lors du setup des test (des jwt éligibles à l'enquête et d'autres non)
* s'assurer que l'email est envoyé pour les jetons éligibles
* s'assurer qu'il n'est pas envoyé pour les jwt non éligibles

Le dernier cas n'est pas testé ici, comme plan de specs je verrai bien quelque chose du genre : 

```ruby 
describe do 
  subject { described_class.call } 

  it { is_expected.to be_success } 

  before do 
    # setup les jetons en base
  end 

  context 'when tokens are eligible to survey' do 
    it 'sends an email to token owners' 
  end 

  context 'when tokens are not eligible to survey' do 
    it 'does not send any email to token owners' 
  end 
end 
```

Reply:
> On a pas besoin de connaitre les rouages internes dans le cadre de ces specs. Si on devait mock une dépendance externe oui mais ce n'est pas le cas en l'occurence.

En bon appliquant du principe de séparation des préoccupations (_separation of concerns_) que je suis, je peux dire que si, il s'agit bien d'une dépendance externe (relative à une logique distincte de celle d'une opération), parce que la logique et le périmètre d'une opération ne sont pas les mêmes que ceux d'un modèle, et que chacun à sa responsabilité. La logique du modèle a été testé unitairement et il ne serait pas DRY de la re-tester ici.

> On pourrait si on voulait résoudre des problèmes de perfs d'exécution des tests : stub le modèle `JwtApiEntreprise` et tester le comportement de l'opération en fonction de son retour.

Plutôt que d'essayer de résoudre des problèmes de perfs d'exécution des tests (que nous n'avons pas), l'idée est de tester le bon fonctionnement de la logique dans le cadre du périmètre de la responsabilité de l'objet (qui est une opération). Pour notamment vérifier que la commande `fetch_eligible_tokens` fait bien ce qu'elle doit faire, sans aller jusqu'à (re-)tester de la logique sous-jacente d'une méthode de classe de modèle (`JwtApiEntreprise.seven_days_ago_created_tokens`, notamment) ; ma proposition c'est d'utiliser l'API publique du modèle (qui doit être testé distinctement), de pouvoir avoir confiance en elle, et de pouvoir s'appuyer sur elle en l'appelant. A ce stade, c'est une primitive en laquelle il serait irrationnel de vouloir douter. Tout comme je n'ai pas besoin de tester les méthodes internes au framework que nous utilisons. Dans le même ordre d'idée, lorsque je teste le job `AccessRequestSatisfactionSurveyJob` qui exploite l'opération `JwtApiEntreprise::Operation::AccessRequestSatisfactionSurvey`, je n'ai pas besoin de (re-)tester l'opération dans le test du comportement du job. J'ai confiance en elle, je sais qu'elle fonctionne parce qu'elle a été testé explicitement, et je peux me concentrer sur le test propre à la logique du job. Ça permet de tester une seule fois le code à chaque étape et de détecter d'éventuels problèmes dès qu'ils arrivent sur une partie du code.

> Et tester aussi `JwtApiEntreprise` en isolation pour s'assurer que la méthode `seven_days_ago_created_tokens` fait bien ce qu'on veut.

Je pense que nous devons avoir ce genre de tests. J'ai d'ailleurs proposé des tests sur cette méthode pour ce modèle.

> Là en l'occurence c'est sur le comportement de l'opération que l'on doit porter attention donc :
> * créer plusieurs jetons en base lors du setup des test (des jwt éligibles à l'enquête et d'autres non)
> * s'assurer que l'email est envoyé pour les jetons éligibles
> * s'assurer qu'il n'est pas envoyé pour les jwt non éligibles

Je veux bien faire ces tests d'intégration. Je n'ai rien contre, au contraire, mais je pense qu'il ne faudrait pas les substituer aux tests unitaires.

Reply:
Ma remarque était purement en terme de *Behaviour Driven Development* qui est la logique de test de tout le répo. 

Ce que l'on teste dans les opération, c'est leur comportement (son contrat fonctionnel, cf les trois points que j'avais partagé dans mon commentaire précédent). Si j'ouvre le code de l'opération je vais le voir tout de suite qu'elle appelle la méthode `seven_days_ago_created_tokens`. Les specs (surtout celles des opérations) je veux qu'elles me renseignent sur leur contrat fonctionnel. 

C'est très bien d'avoir des tests unitaires pour `seven_days_ago_created_tokens` dans le modèle 👍🏼 ils sont nécessaires. Par contre il n'est pas nécessaire de tester les rouages internes de l'opération. Au contraire je trouve ça dommage : si demain on change la façon de faire pour la sélection des jetons (ne serait qu'un renaming de la méthode) cela casse les specs alors que le contrat de l'opération est toujours respecté et imo c'est un problème. Dans le cadre de tests unitaires on peut vouloir tester que telle ou telle méthode a bien été appelée, comme ici la logique de tests est orientée BDD je le lis ça comme "l'opération doit appeler la méthode `seven_days_ago_created_tokens` ce qui ne me renseigne en rien sur son contrat vis à vis de son comportement.

Ce n'est pas critique, ce qui est critique ce sont les tests qui manquaient pour s'assurer que l'opération fait bien ce que l'on veut fonctionnellement. A côté, les tests qui s'assurent que l'opération appelle les bonnes méthodes en interne c'est du bonus mais en aucun cas ça ne les remplace. 

Petite aparté sur le BDD : d'expérience de telles suites de tests sont en général plus claires et robustes. Elles peuvent être utilisées comme documentation métier (et je les utilise pour ça) car elles décrivent le comportement métier du code.  Ca a plus de valeur de tester que le code fait bien ce qu'on attend de lui fonctionnellement plutôt que de s'assurer que "la pelote de laine" de calls et de sub-call entre tous les composants est respectée. Et c'est plus robuste car dans ce dernier cas il faut refactor toutes les specs en cas de refactor pour s'assurer que la nouvelle chaine d'appels est respectée et c'est comme ça que des tests de régression apparaissent. Avec une suite de test BDD au vert qui n'est pas couplée au rouages internes de ses dépendances tu as l'assurance de ne casser le contrat fonctionnel tant qu'elle reste au vert, et ce même dans le cadre d'un refactor. 

Bref, BDD++ et si tu veux garder cette spec en bonus feel free to do so. Je voulais surtout attirer ton attention sur la philosophie d'écriture des tests.

Reply:
Je suis d'accord avec toi. Je vais modifier le test de l'opération. Ce ne sera du coup plus un test unitaire mais un test d'intégration, avec quelques duplications entre les tests de comportement du modèle et ceux de l'opération. Mais ces duplications sont saines, IMO.